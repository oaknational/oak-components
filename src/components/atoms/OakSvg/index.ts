export { OakSvg } from "./OakSvg";
export type { OakSvgProps, OakSvgNames } from "./OakSvg";
