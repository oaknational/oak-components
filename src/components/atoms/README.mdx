# Atoms

## Rules for components in this folder

- Can import other atoms
- Imports from molecules and organisms are forbidden!
- Generic
- Unstyled
- Avoid embedded logic and state
- No sub-components
  (eg. Box)
