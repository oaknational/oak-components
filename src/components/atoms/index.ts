export * from "./OakBox";
export * from "./OakFlex";
export * from "./OakSpan";
export * from "./OakForm";
export * from "./OakImage";
export * from "./OakTypography";
export * from "./OakHeading";
export * from "./OakP";
export * from "./OakUL";
export * from "./OakOL";
export * from "./OakLI";
export * from "./OakLabel";
export * from "./OakIcon";
export * from "./OakGrid";
export * from "./OakGridArea";
export * from "./OakThemeProvider";
export * from "./OakMaxWidth";
export * from "./OakCloudinaryImage";
export * from "./OakKbd";
export * from "./OakGlobalStyle";
export * from "./OakAnchorTarget";
export * from "./OakFieldSet";
