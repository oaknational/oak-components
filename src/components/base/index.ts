export * from "./OakBox";
export * from "./OakFlex";
export * from "./OakBaseCard";
export * from "./OakSpan";
export * from "./OakForm";
export * from "./OakImage";
export * from "./OakTypography";
export * from "./OakHeading";
export * from "./OakP";
export * from "./OakUL";
export * from "./OakOL";
export * from "./OakLI";
export * from "./OakLabel";
