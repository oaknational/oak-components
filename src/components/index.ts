export * from "./base";
// export * from "./ui"; // uncomment this line once there are components in this folder
// export * from "./integrated"; // uncomment this line once there are components in this folder
