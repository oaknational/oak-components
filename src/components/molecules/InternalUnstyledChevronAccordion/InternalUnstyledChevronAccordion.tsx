import React, { ReactNode } from "react";
import styled from "styled-components";

import {
  OakBoxProps,
  OakFlex,
  OakIcon,
  OakBox,
  oakBoxCss,
} from "@/components/atoms";
import {
  InternalAccordionButton,
  InternalAccordionContent,
} from "@/components/atoms/InternalAccordion";
import useAccordionContext from "@/components/atoms/InternalAccordion/useAccordionContext";
import InternalAccordionProvider from "@/components/atoms/InternalAccordion/InternalAccordionProvider";
import { flexStyle, FlexStyleProps } from "@/styles/utils/flexStyle";
import { ColorStyleProps } from "@/styles/utils/colorStyle";
import { parseDropShadow } from "@/styles/helpers/parseDropShadow";
import { parseColorFilter } from "@/styles/helpers/parseColorFilter";

export type InternalUnstyledChevronAccordionProps = {
  /**
   * The header of the accordion
   */
  header: ReactNode;
  /**
   * Whether the accordion should be open initially
   */
  initialOpen?: boolean;
  /**
   * The content of the accordion
   */
  content: ReactNode;
  /**
   * Optional subheader to display above the fold
   */
  subheader?: ReactNode;
  /**
   * The id of the accordion
   */
  id: string;
  /**
   * Aria label for the button when the accordion is open
   */
  ariaLabelOpen?: string;
  /**
   * Aria label for the button when the accordion is closed
   */
  ariaLabelClose?: string;
} & FlexStyleProps &
  OakBoxProps &
  ColorStyleProps;

const StyledAccordionButton = styled(InternalAccordionButton)<FlexStyleProps>`
  ${flexStyle}
  ${oakBoxCss}
  &:hover {
    .chevron-icon {
      filter: ${parseColorFilter("grey60")};
    }
  }
  &:focus-visible {
    .focus-outline {
      box-shadow: ${parseDropShadow("drop-shadow-centered-lemon")},
        ${parseDropShadow("drop-shadow-centered-grey")};
    }
  }
`;

/**
 * An accordion component that can be used to show/hide content
 */

const Accordion = ({
  header,
  content,
  id,
  subheader,
  ariaLabelOpen = "Close accordion",
  ariaLabelClose = "Open accordion",
  ...styleProps
}: InternalUnstyledChevronAccordionProps) => {
  const { isOpen } = useAccordionContext();

  return (
    <OakFlex $pv={"inner-padding-s"} $flexDirection={"column"} {...styleProps}>
      <OakFlex $justifyContent={"space-between"} $alignItems={"center"}>
        {header}
        <OakBox>
          <StyledAccordionButton
            id={id}
            aria-label={isOpen ? ariaLabelOpen : ariaLabelClose}
          >
            <OakBox
              className="focus-outline"
              $borderRadius={"border-radius-s"}
              $mr={"space-between-xs"}
            >
              <OakIcon
                className="chevron-icon"
                iconName="chevron-down"
                $width="all-spacing-7"
                $height="all-spacing-7"
                alt="An arrow to indicate whether the item is open or closed"
                style={{
                  transform: isOpen ? "rotate(180deg)" : "none",
                  transition: "all 0.3s ease 0s",
                }}
              />
            </OakBox>
          </StyledAccordionButton>
        </OakBox>
      </OakFlex>
      {subheader}
      <OakBox $position={"relative"} $overflow={"auto"}>
        <InternalAccordionContent aria-labelledby={id}>
          {content}
        </InternalAccordionContent>
      </OakBox>
    </OakFlex>
  );
};

/**
 * - InternalUnstyledChevronAccordion has a chevron icon that rotates when the accordion is open.
 * - Unlike InternalChevronAccordion, it has no border effects for hover or focus states.
 * - Only the chevron is interactive so as to allow interactive elements to be placed in the header.
 * - The intention is for these to be added by consuming components as needed.
 */

export const InternalUnstyledChevronAccordion = ({
  initialOpen = false,
  ...props
}: InternalUnstyledChevronAccordionProps) => {
  return (
    <InternalAccordionProvider isInitialOpen={initialOpen}>
      <Accordion initialOpen {...props} />
    </InternalAccordionProvider>
  );
};
