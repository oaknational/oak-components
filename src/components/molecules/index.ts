export * from "./OakFieldError";
export * from "./OakLoadingSpinner";
export * from "./OakPrimaryButton";
export * from "./OakPrimaryInvertedButton";
export * from "./OakSecondaryButton";
export * from "./OakTertiaryButton";
export * from "./OakRadioButton";
export * from "./OakRadioGroup";
export * from "./OakCheckBox";
export * from "./OakTextInput";
export * from "./OakRoundIcon";
export * from "./OakTooltip";
export * from "./OakBackLink";
export * from "./OakBulletList";
export * from "./OakHandDrawnCard";
export * from "./OakHandDrawnCardWithIcon";
export * from "./OakHandDrawnHR";
export * from "./OakPromoTag";
export * from "./OakLessonInfoCard";
export * from "./OakCollapsibleContent";
export * from "./OakLink";
export * from "./OakSecondaryLink";
export * from "./OakDragAndDropInstructions";
export * from "./OakDraggable";
export * from "./OakDroppable";
export * from "./OakDraggableFeedback";
export * from "./OakOutlineAccordion";
export * from "./OakAccordion";
export * from "./OakModal";
export * from "./OakModalCenter";
export * from "./OakCardWithHandDrawnBorder";
export * from "./OakSecondaryButtonAsRadio";
export * from "./OakButtonAsRadioGroup";
