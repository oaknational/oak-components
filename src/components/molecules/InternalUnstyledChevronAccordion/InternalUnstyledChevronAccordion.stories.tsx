import React from "react";
import { Meta, StoryObj } from "@storybook/react";

import { InternalUnstyledChevronAccordion } from "./InternalUnstyledChevronAccordion";

import { OakSmallSecondaryButtonWithDropdown } from "@/components/molecules/OakSmallSecondaryButtonWithDropdown";
import { OakBox, OakP } from "@/components/atoms";

const meta: Meta<typeof InternalUnstyledChevronAccordion> = {
  component: InternalUnstyledChevronAccordion,
  tags: ["autodocs"],
  parameters: {
    controls: {
      include: ["header", "headerAfterSlot", "children"],
    },
  },
  argTypes: {
    children: {
      control: {
        type: "text",
      },
    },
    header: {
      control: {
        type: "text",
      },
    },
  },
  args: {
    id: "accordion-1",
    header: <OakBox>Header goes here</OakBox>,
    children:
      "Any cookies required for video or other embedded learning content to work",
  },
  render: (args) => <InternalUnstyledChevronAccordion {...args} />,
};
export default meta;

type Story = StoryObj<typeof InternalUnstyledChevronAccordion>;

export const Default: Story = {};

export const MultilineHeader: Story = {
  args: {
    header: (
      <OakBox $background={"amber50"} $textAlign={"left"}>
        <OakP>Header goes here</OakP>
        <OakP>This is a multiline header</OakP>
      </OakBox>
    ),
  },
};

export const InputAsHeader: Story = {
  args: {
    header: (
      <OakBox $background={"amber50"} $textAlign={"left"}>
        <OakSmallSecondaryButtonWithDropdown primaryActionText="Click me">
          <OakSmallSecondaryButtonWithDropdown.Item>
            Action 1
          </OakSmallSecondaryButtonWithDropdown.Item>
          <OakSmallSecondaryButtonWithDropdown.Item>
            Action 2
          </OakSmallSecondaryButtonWithDropdown.Item>
          <OakSmallSecondaryButtonWithDropdown.Item>
            Action 3
          </OakSmallSecondaryButtonWithDropdown.Item>
        </OakSmallSecondaryButtonWithDropdown>
      </OakBox>
    ),
  },
};
