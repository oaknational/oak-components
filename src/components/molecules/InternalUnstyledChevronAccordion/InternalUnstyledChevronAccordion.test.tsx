import React from "react";
import { create } from "react-test-renderer";
import { act, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";

import { InternalUnstyledChevronAccordion } from "./InternalUnstyledChevronAccordion";

import { OakThemeProvider } from "@/components/atoms";
import { oakDefaultTheme } from "@/styles";
import renderWithTheme from "@/test-helpers/renderWithTheme";

describe(InternalUnstyledChevronAccordion, () => {
  it("matches snapshot", () => {
    const tree = create(
      <OakThemeProvider theme={oakDefaultTheme}>
        <InternalUnstyledChevronAccordion
          initialOpen
          header="See more"
          content={<div>Here it is</div>}
          id="see-more"
        />
      </OakThemeProvider>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("toggles open and closed", () => {
    const { queryByText, queryByRole, getByRole } = renderWithTheme(
      <InternalUnstyledChevronAccordion
        initialOpen={false}
        header="See more"
        content={<div>Here it is</div>}
        id="see-more"
      />,
    );

    // expect(getByRole("region")).not.toBeVisible();

    act(() => {
      fireEvent.click(getByRole("button"));
    });

    expect(getByRole("region")).toBeVisible();

    expect(queryByText("Here it is")).toBeInTheDocument();

    act(() => {
      fireEvent.click(getByRole("button"));
    });

    expect(queryByRole("region")).toBeNull();
    expect(queryByText("Here it is")).not.toBeVisible();

    act(() => {
      fireEvent.click(getByRole("button"));
    });

    expect(getByRole("region")).toBeInTheDocument();
  });

  it("changes aria-label based on open state", () => {
    const { getByRole } = renderWithTheme(
      <InternalUnstyledChevronAccordion
        initialOpen={false}
        header="See more"
        content={<div>Here it is</div>}
        id="see-more"
        ariaLabelOpen="Close accordion"
        ariaLabelClose="Open accordion"
      />,
    );

    expect(getByRole("button")).toHaveAttribute("aria-label", "Open accordion");

    act(() => {
      fireEvent.click(getByRole("button"));
    });

    expect(getByRole("button")).toHaveAttribute(
      "aria-label",
      "Close accordion",
    );
  });
});
