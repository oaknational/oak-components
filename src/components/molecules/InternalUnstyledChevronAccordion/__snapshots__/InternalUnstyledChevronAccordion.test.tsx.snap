// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InternalUnstyledChevronAccordion matches snapshot 1`] = `
.c0 {
  padding-top: 0.75rem;
  padding-bottom: 0.75rem;
  font-family: --var(google-font),Lexend,sans-serif;
}

.c2 {
  font-family: --var(google-font),Lexend,sans-serif;
}

.c4 {
  font-family: --var(google-font),Lexend,sans-serif;
}

.c4:hover {
  cursor: pointer;
}

.c8 {
  margin-right: 0.75rem;
  border-radius: 0.25rem;
  font-family: --var(google-font),Lexend,sans-serif;
}

.c9 {
  position: relative;
  width: 2rem;
  min-width: 2rem;
  height: 2rem;
  min-height: 2rem;
  font-family: --var(google-font),Lexend,sans-serif;
}

.c11 {
  position: relative;
  overflow: auto;
  font-family: --var(google-font),Lexend,sans-serif;
  overflow: auto;
}

.c10 {
  object-fit: contain;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c6 {
  font: inherit;
  color: inherit;
  border: none;
  background: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  outline: none;
  padding: 0;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-family: --var(google-font),Lexend,sans-serif;
}

.c7:hover .chevron-icon {
  -webkit-filter: invert(34%) sepia(0%) saturate(698%) hue-rotate(158deg) brightness(95%) contrast(89%);
  filter: invert(34%) sepia(0%) saturate(698%) hue-rotate(158deg) brightness(95%) contrast(89%);
}

.c7:focus-visible .focus-outline {
  box-shadow: 0 0 0 0.125rem rgba(255,229,85,100%), 0 0 0 0.3rem rgba(87,87,87,100%);
}

<div>
  <div
    class="c0 c1"
  >
    <div
      class="c2 c3"
    >
      See more
      <div
        class="c2"
      >
        <button
          aria-expanded="true"
          aria-label="Close accordion"
          class="c4 c5 c6  c7"
          id="see-more"
          type="button"
        >
          <div
            class="c8 focus-outline"
          >
            <div
              class="c9 chevron-icon"
              style="transform: rotate(180deg); transition: all 0.3s ease 0s;"
            >
              <img
                alt="An arrow to indicate whether the item is open or closed"
                class="c10"
                data-nimg="fill"
                decoding="async"
                loading="lazy"
                src="https://res.cloudinary.com/mock-cloudinary-cloud/image/upload/v1699953557/icons/botfld6brychmttwtv6u.svg"
                style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
              />
            </div>
          </div>
        </button>
      </div>
    </div>
    <div
      class="c11"
    >
      <div
        aria-labelledby="see-more"
        class="c2"
        role="region"
      >
        <div>
          Here it is
        </div>
      </div>
    </div>
  </div>
</div>
`;
